hellodagger/src/androidTest/java/com/airbnb/mvrx/helloDagger/HelloFragmentTest.kt

```diff
@@ -0,0 +1,19 @@
+package com.airbnb.mvrx.helloDagger
+
+import androidx.fragment.app.testing.launchFragmentInContainer
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.*
+import org.hamcrest.CoreMatchers.not
+import org.junit.Test
+
+class HelloFragmentTest {
+
+    @Test
+    fun showsLoadingMessageWhenCreated() {
+        val scenario = launchFragmentInContainer<HelloFragment>()
```

Comment:
I'm surprised this works :O `launchFragmentInContainer` should launch an empty activity hosting your fragment, this activity which is not a `MvRxActivity` which is a hard requirement from the view model factory setup if I understand the code correctly.

Reply:
This test does indeed pass, and upon further investigation I think this is because of the lazy nature of 'by fragmentViewModels()` delegate. I think `invalidate()` is never called, and that's why the ViewModel is never accessed within the fragment. Since it is lazily initialized, it also means that it is never created, and thus the underlying activity is never used for its creation using the multibinding map.

Thanks for pointing this out. I'm replacing the usage of `FragmentScenario` with `ActivityScenario` and re writing the test using it.

Reply:
I'm running into a problem here. This is what my test currently looks like:

```kotlin
@Test
    fun fragmentIsInLoadingStateWhenCreated() {
        val scenario: ActivityScenario<MainActivity> = ActivityScenario.launch(MainActivity::class.java)

        scenario.onActivity { activity: MainActivity ->
            val fragment = activity.findFragmentById<HelloFragment>(R.id.rootContainer)

            withState(fragment.viewModel) {
                assert(it.message is Loading)
            }

            onView(withId(R.id.messageTextView)).check(matches(withText(R.string.hello_fragment_loading_text)))
            onView(withId(R.id.helloButton)).check(matches(not(isEnabled())))
        }
    }
```

The problem is that the test just never completes. I'm not sure what's the problem here. On adding some log statements, I can see that the call to the first `onView(...)` never completes.

Is there anything I'm doing wrong here?

Reply:
It's not exactly lazy. It's a "lifecycle away aware lazy" and auto instantiates itself in onStart. `postInvalidate()` is also called in onStart() so that MvRx works with static screens and for cases like this.